src/buffer.h

```diff
@@ -29,89 +29,45 @@
 #include "zim_types.h"
 #include "endian_tools.h"
 #include "debug.h"
+#include <zim/blob.h>
 
 namespace zim {
 
-class Buffer : public std::enable_shared_from_this<Buffer> {
-  public:
-    explicit Buffer(zsize_t size)
-      : size_(size)
-    {
-      ASSERT(size_.v, <, SIZE_MAX);
-    };
+class Buffer {
+  public: // types
+    typedef std::shared_ptr<const char> DataPtr;
 
-    Buffer(const Buffer& ) = delete;
-    void operator=(const Buffer& ) = delete;
+  public: // functions
+    static const Buffer makeBuffer(const char* data, zsize_t size);
+    static const Buffer makeBuffer(const DataPtr& data, zsize_t size);
+    static Buffer makeBuffer(zsize_t size);
 
-    virtual ~Buffer() {};
-    const char* data(offset_t offset=offset_t(0)) const {
-      ASSERT(offset.v, <=, size_.v);
-      return dataImpl(offset);
-    }
+    const char* data(offset_t offset=offset_t(0)) const;
+    char* data(offset_t offset=offset_t(0));
 
     char at(offset_t offset) const {
-        return *(data(offset));
+      return *(data(offset));
     }
-    zsize_t size() const { return size_; }
-    std::shared_ptr<const Buffer> sub_buffer(offset_t offset, zsize_t size) const;
+    zsize_t size() const { return m_size; }
+    const Buffer sub_buffer(offset_t offset, zsize_t size) const;
 
     template<typename T>
     T as(offset_t offset) const {
-      ASSERT(offset.v, <, size_.v);
-      ASSERT(offset.v+sizeof(T), <=, size_.v);
+      ASSERT(offset.v, <, m_size.v);
+      ASSERT(offset.v+sizeof(T), <=, m_size.v);
       return fromLittleEndian<T>(data(offset));
     }
```

Comment:
Is this still needed?

Reply:
They are securities to be sure we don't do bufferoverflow. Why should we remove them ?

Reply:
I was asking about the function `Buffer::as<T>()`

Reply:
Indeed, I've removed it.